dgl_sparse/src/sddmm.cc

```diff
@@ -0,0 +1,130 @@
+/**
+ *  Copyright (c) 2022 by Contributors
+ * @file sddmm.cc
+ * @brief DGL C++ sparse SDDMM operator implementation.
+ */
+// clang-format off
+#include <sparse/dgl_headers.h>
+// clang-format on
+
+#include <sparse/sparse_matrix.h>
+#include <sparse/spmm.h>
+#include <torch/script.h>
+
+#include "./utils.h"
+
+namespace dgl {
+namespace sparse {
+
+using namespace torch::autograd;
+
+class SDDMMAutoGrad : public Function<SDDMMAutoGrad> {
+ public:
+  static torch::Tensor forward(
+      AutogradContext* ctx, const c10::intrusive_ptr<SparseMatrix>& sparse_mat,
+      torch::Tensor mat1, torch::Tensor mat2_tr);
+
+  static tensor_list backward(AutogradContext* ctx, tensor_list grad_outputs);
+};
+
+void _SDDMMSanityCheck(
+    const c10::intrusive_ptr<SparseMatrix>& sparse_mat, torch::Tensor mat1,
+    torch::Tensor mat2) {
+  CHECK_EQ(mat1.dim(), 2)
+      << "SDDMM: the first dense matrix should be two-dimensional.";
+  CHECK_EQ(mat2.dim(), 2)
+      << "SDDMM: the second dense matrix should be two-dimensional.";
+  CHECK_EQ(sparse_mat->shape()[0], mat1.size(0))
+      << "SDDMM: the first dense matrix should have the same first dimension "
+         "as the sparse matrix";
+  CHECK_EQ(sparse_mat->shape()[1], mat2.size(1))
+      << "SDDMM: the seond dense matrix should have the same second dimension "
+         "as the sparse matrix";
+  CHECK_EQ(mat1.size(1), mat2.size(0))
+      << "SDDMM: the second dimension of the first dense matrix should be "
+         "equal to the first dimension of the second dense matrix.";
+  CHECK_EQ(mat1.dtype(), mat2.dtype())
+      << "SDDMM: the two dense matrices should have the same dtype.";
+}
+
+torch::Tensor SDDMMImpl(
+    const c10::intrusive_ptr<SparseMatrix>& sparse_mat, torch::Tensor mat1,
+    torch::Tensor mat2_tr) {
+  const int64_t out_row = sparse_mat->nnz();
+  const std::vector<int64_t> shape({out_row});
+  auto ret = torch::zeros(shape, mat1.options());
+  const std::string op = "dot";
+  auto dgl_mat1 = TorchTensorToDGLArray(mat1);
+  auto dgl_mat2_tr = TorchTensorToDGLArray(mat2_tr);
+  auto dgl_ret = TorchTensorToDGLArray(ret);
+  // Use CSR if the sparse matrix has CSR or does not have COO. Otherwise use
+  // COO.
+  if (sparse_mat->HasCSR() || !sparse_mat->HasCOO()) {
```

Comment:
Will this crash if the !sparse_mat->HasCSR() && !sparse_mat->HasCOO()?

Reply:
If the sparse_mat should either HasCSR() or  HasCOO(). Then do:
If (sparse_mat->HasCSR()) {
} else if (sparse_mat->HasCOO()) {
} else {
    throw ...
}

Reply:
> Will this crash if the !sparse_mat->HasCSR() && !sparse_mat->HasCOO()?

It would not crash because it will convert CSC to CSR for computation. The logic here is to use CSR or COO if they exist. Otherwise convert CSC to CSR for computation.